### PR TITLE
[PVR] Global direct channel number input & global unique channel numbers

### PR DESCRIPTION
**Pull request by ksooo**
First commit implements a feature that was requested in the forum and that makes Kodi behave like other settop boxes: if direct channel number input is used to switch channels and the currently active channel group does not contain a channel with the requested number, PVR tries to obtain the channel with the entered number from the "All channels" group, sets this group active and then switches to the requested channel.

Second commit also implements a feature common to other settop boxes - to use the same channel number for a channel  in all channel groups where this channel appears:

screenshot002
screenshot003

These changes are runtime-tested on macOS, latest Kodi master.

@xhaggi mind doing the code review?

### CODE REVIEW

**Comment by xhaggi** [comment]
I tend to drop the group specific channel number and use the unique number instead. I don't know any TV or settop box that do it like we do. That way we can search for the channel by number over all active groups instead of just fallback to the "All channels" group. For me the "All channels" group is hidden, because all my channels are organized in groups and I don't want the "All channels" group to appear.

**Comment by ksooo** [comment]
> I tend to drop the group specific channel number and use the unique number instead. I don't know any TV or settop box that do it like we do.

100% ack. I will drop the setting and switch logic to use unique numbers.

>  That way we can search for the channel by number over all active groups instead of just fallback to the "All channels" group. For me the "All channels" group is hidden, because all my channels are organized in groups and I don't want the "All channels" group to appear.

 That's not actually dependent on the unique numbers feature, but a very valid point. I had not in mind that channels from the all group can be hidden and wil change the PR accordingly.

**Comment by xhaggi** [comment]
Thanks .. let me know If you want me to take another look

**Comment by ksooo** [comment]
@xhaggi updated the PR, like we discussed.

File: xbmc/pvr/PVRGUIActions.cpp [hunk 1]
```diff
@@ -1685,8 +1685,28 @@ namespace PVR
...
+
+          if (!channel || !channel->HasPVRChannelInfoTag())
+          {
+            // channel number present in any group?
+            const CPVRChannelGroups* groupAccess = CServiceBroker::GetPVRManager().ChannelGroups()->Get(bRadio);
+            const std::vector<CPVRChannelGroupPtr> groups = groupAccess->GetMembers(true);
+            for (const auto &currentGroup : groups)
+            {
+              channel = currentGroup->GetByChannelNumber(channelNumber);
+              if (channel)
```
Comment (xhaggi) [inline xbmc/pvr/PVRGUIActions.cpp]: should we also need to check for `channel->HasPVRChannelInfoTag()` or could we drop the check for `channel->HasPVRChannelInfoTag()`

File: xbmc/pvr/PVRGUIActions.cpp [hunk 2]
```diff
@@ -1685,8 +1685,28 @@ namespace PVR
...
+          bool bRadio = playingChannel->IsRadio();
+          const CPVRChannelGroupPtr group = CServiceBroker::GetPVRManager().GetPlayingGroup(bRadio);
+          CFileItemPtr channel = group->GetByChannelNumber(channelNumber);
+
+          if (!channel || !channel->HasPVRChannelInfoTag())
+          {
+            // channel number present in any group?
+            const CPVRChannelGroups* groupAccess = CServiceBroker::GetPVRManager().ChannelGroups()->Get(bRadio);
+            const std::vector<CPVRChannelGroupPtr> groups = groupAccess->GetMembers(true);
+            for (const auto &currentGroup : groups)
```
Comment (xhaggi) [inline xbmc/pvr/PVRGUIActions.cpp]: use `groupAccess->GetMembers(true)` directly here to save the declaration.

**Review by xhaggi — APPROVED** [review]
Looks good apart from the minors

File: xbmc/pvr/PVRGUIActions.cpp (hunk 1 above)
Comment (ksooo) [reply xbmc/pvr/PVRGUIActions.cpp]: Fixed. See 3rd commit. No need to check for a channel tag at all now.

File: xbmc/pvr/PVRGUIActions.cpp (hunk 2 above)
Comment (ksooo) [reply xbmc/pvr/PVRGUIActions.cpp]: I did that because I hate loooooooooong code lines.

File: xbmc/pvr/PVRGUIActions.cpp [hunk 3]
```diff
@@ -1685,9 +1685,29 @@ namespace PVR
...
+          bool bRadio = playingChannel->IsRadio();
+          const CPVRChannelGroupPtr group = CServiceBroker::GetPVRManager().GetPlayingGroup(bRadio);
+          CFileItemPtr channel = group->GetByChannelNumber(channelNumber);
+
+          if (!channel)
+          {
+            // channel number present in any group?
+            const CPVRChannelGroups* groupAccess = CServiceBroker::GetPVRManager().ChannelGroups()->Get(bRadio);
+            const std::vector<CPVRChannelGroupPtr> groups = groupAccess->GetMembers(true);
+            for (const auto &currentGroup : groups)
```
Comment (xhaggi) [inline xbmc/pvr/PVRGUIActions.cpp]: Ok

**Comment by ksooo** [comment]
jenkins build this please

**Comment by ksooo** [comment]
@Rechi any idea why jenkins is acting up? Anything related to this PR?

**Comment by ksooo** [comment]
jenkins build this please

**Comment by Rechi** [comment]
reason for failing jenkins is #13112

**Pull request merged**